*: framework agnostic loaders

tharvik opened the pull request:
on the road of supporting ONNX, we first need to extract TFJS from the code base, making discojs framework agnostic.
based on #735.

* adding a Dataset type, wrapping standard [AsyncGenerator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AsyncGenerator) with some helpers (map, zip, split, batch, ...)
  * for now, `disco.fit`, `validator.{assess,predict}`, requires more type informations (via `TypedDataset`) as theses are supporting any dataset for any task. it should be enforced at some point via `Task` splitting so that we can drop it.
  * now, any users of disco have to use it, keeping the remaining `Datasplit`, `Data`, `tf.*` internal to disco (and removing theses when preprocessing and task typing are reworked)
* also exposes types: `Image`, `Tabular` & `Text` 
* redo loaders to be small functions rather than cumbersome-to-use classes
* added some `convertors` as a potential new way to do preprocessing
  * flat functions, that encodes type changes instead of wrapping in `tf.TensorContainer` (pretty much equivalent to `any`)
* webapp are now typing its dataset, having a clear separation between Image, Tabular & Text, with `Data` component emitting changes

tharvik committed:
discojs/dataset: revamp

tharvik committed:
webapp/data: rework

tharvik committed:
discojs-*: rework loaders

tharvik committed:
*: use Dataset

tharvik committed:
discojs/convertors: initial

tharvik committed:
discojs*: rm *Loader & DatasetBuilder

JulienVig commented on cli/src/benchmark_gpt.ts:
```diff
@@ -72,8 +88,28 @@ async function main(args: Required<CLIArguments>): Promise<void> {
     // to make sure the dataset is batched and tokenized correctly
     task.trainingInformation.batchSize = batchSize
     task.trainingInformation.maxSequenceLength = contextLength
-    const dataset = await loadWikitextData(task)
-    const preprocessedDataset = dataset.train.preprocess().batch().dataset
+    const dataset = loadText('../datasets/wikitext/wiki.train.tokens')
+
+    const maxLength = task.trainingInformation.maxSequenceLength ?? (tokenizer.model_max_length as number) + 1
+    const preprocessedDataset = intoTFGenerator(
+      dataset
+        .map((line) => tokenize_and_left_pad(line, tokenizer, maxLength))
+        .batch(batchSize)
+        .map((batch) => ({
+          xs: tf.tensor2d(batch.map((tokens) => tokens.slice(0, -1)).toArray()),
+          ys: tf.stack(
+            batch
+              .map(
+                (tokens) =>
+                  tf.oneHot(
+                    tokens.slice(1),
+                    tokenizer.model.vocab.length + 1,
+                  ) as tf.Tensor2D,
+              )
+              .toArray(),
+          ) as tf.Tensor3D,
+        })),
+    );
```
Is the Task's `preprocessingFunctions` attribute obsolete now? I guess they are still used by the webapp on default tasks
Why reimplement the preprocessing rather than use wikitext's preprocessingFunctions ?
I feel a bit uneasy with this way of preprocessing text dataset because it requires users a deep understand of the implementation of tokenize_and_left_pad to know that they should then slice `xs` and `ys` accordingly.

It also seems that setting `task.trainingInformation.batchSize` and `maxSequenceLength` is useless now

tharvik replied:
the issue with `preprocessingFunctions` is that there aren't typesafe at all. there is no way that I can think of in TS to transform values by applying an array of function that doesn't use a common type (in this case, `tf.TensorContainer` which is akin to `any`, that gets cast to the correct shape in every `apply`).
what I've in mind is that `Disco` can take "user-friendly Dataset" (`Dataset<Image/Tabular/Text>`) and takes care of shaping theses into what `models.*` can handle (tokenized inputs, resized images, ...). as such, users that direclty want to use models should apply theses functions themself.
but yeah, as this is a ongoing update, this is quite ugly to use models direclty for now. in a next iteration, models will directly take our `Dataset` and takes care of mapping to its own tensors (which would be framework specific).
does that makes sense and/or do you think that's a good idea?

JulienVig replied:
Yes this makes sense! I was mostly wondering about the code duplciation, I agree that the current preprocessing should be reworked and how you describe it sounds good. I would try to follow pytorch and tensorflow's api as much as we can

JulienVig commented on discojs-node/src/loaders/index.ts:
```diff
@@ -0,0 +1,6 @@
+export { load as loadCSV } from "./csv.js";
+export {
+  load as loadImage,
+  loadAllInDir as loadImagesInDir,
```
Why not directly rename `loadAllInDir` to `loadImagesInDir`?

tharvik replied:
I'm overly scope sensitive: names inside a file called `image.ts` doesn't need to contain `Image` as it is already in the filename. this has some logic when moving files around, as it avoid changing the file itself, only its parents.

JulienVig commented on discojs-node/src/loaders.spec.ts:
```diff
@@ -0,0 +1,68 @@
+}
+
+describe("csv parser", () => {
+  it("parses basic file", async () => {
```
Could you add a test for a csv with missing values? (e.g. titanic has missing values)

tharvik replied:
what do you mean by missing value? the loader is quite basic, only checking that is has the same number of fields as the headers. for example
```
a,b,c
1,2,3
4,,6
```
will get parsed as
```js
[
  {a: "1", b: "2", c: "3"},
  {a: "4", b: "", c: "6"},
]
```

I think it should stay dumb and that convertors should ensure that theses fields are valid (ie, in a specific range or element of a list, mapped to number, ...).

JulienVig replied:
Oops forgot to answer. Yes the b in your example was what I was talking about so all good!

JulienVig commented on discojs-node/src/loaders/image.ts:
```diff
@@ -0,0 +1,26 @@
+export async function load(path: string): Promise<Image> {
+  const { data, info } = await sharp(path).removeAlpha().raw().toBuffer({
+    resolveWithObject: true,
+  });
```
I used to have an issue with the default number of channel being 1 for B&W images, is it always 3 with sharp now?

tharvik replied:
I rewrote part of it, now `Image` is a real class (for checking the size of the array with width, height and depth) and type-encode most arguments. I'm failling with 1 & 2 channels image now, is it common to have B&W images?

JulienVig replied:
Medical images are frequently black and white yes

tharvik replied:
alright, added support for depth of one, and related convertor for expanding an image to multicolor, thanks for point that out!

JulienVig commented on cli/src/data.ts:
```diff
@@ -1,76 +1,53 @@
+      return ["image", await loadSimpleFaceData()];
+    case "titanic":
+      return ["tabular", loadCSV("../datasets/titanic_train.csv")];
+    case "cifar10":
+      return [
+        "image",
+        (await loadImagesInDir(path.join("..", "datasets", "CIFAR10"))).zip(
+          Repeat("cat"),
+        ),
+      ];
```
There seems to be something wrong with training CIFAR10 with the CLI now.
Previously, running `npm -w cli start -- -u 2 -e 3 -t cifar10` would show an initial loss around 2 (which is slightly higher than expected for a 10-class classification: -log(1/n_classes) = 1) but now it is at 20 which is way too high. I tried fiddling with things but I didn't find the source of the issue, maybe the image loader?

tharvik replied:
ha ho, oupsi, it seems that I inverted the generator oneHot tensor, returning an array full of ones and a single zero for the label, rather than the other way around. hard to find, thanks for the check!

JulienVig commented on discojs-web/src/loaders.spec.ts:
```diff
@@ -0,0 +1,22 @@
+  return ret;
+}
+
+describe("csv parser", () => {
```
Is it hard to add tests for images and text? I know there wasn't any test at all before

tharvik replied:
yeah, for images, it's hard: the testing environnement (jsdom) isn't a full browser and doesn't have support for either OffScreenCanvas nor `URL.createObjectURL`, which is used by the current implementation. I don't see another (more testable and not adding deps) to do so sadly.

for text, there isn't even a loader for now, but that would be easy IMO. I didn't wanted to write one before it was needed as to have a better understanding of the needed interface.

JulienVig commented on discojs/src/convertors.ts:
```diff
@@ -0,0 +1,69 @@
+ * @param length number of tokens
+ * @returns encoded string in an array of token, size of max_length
+ **/
+export function tokenize_and_left_pad(
```
Is there a plan to remove the code duplication between this function and the text preprocessing in the future?

tharvik replied:
yeah, for sure, removing `{Text,Image,Tabular}Processing` and simply use the convertors

JulienVig commented on discojs/src/convertors.ts:
What's the idea of this file? Is it a repository of preprocessing functions?

tharvik replied:
pretty much yeah, it's more than preprocessing as some will be used for postprocessing (such as untokenization, `number -> label` translation, ...)
it has the advantage of not being weird classes but simple function, encoding real type transformation, and fittin pretty nicely into `Dataset.map` (even more nicely if curring was supported in JS but yeah, it's not Haskell)
I'll probably change it some more, probably splitting it into various files, but the interface should be quite similar

that's a proposition of replacement anyway, if you think of another nicer way to do it, do tell me :)

JulienVig replied:
Edit: this is a reply to a comment in the previous review that has been detached from its thread

I like the idea of having simple stateless functions for preprocessing. I would prefer if it was more organized, maybe by dataset type but that can be handled later on as we add more functions. 
I think there may be some overlap between this file and `data/helpers.ts`, could you add some doc about the scope of each file to clarify where functions should go? 
On that matter maybe `convertors.ts` is a confusing name as not all functions there convert things (and some functions in helpers.ts also convert things). What about something like `processing.ts` (for both pre- and post-processing)?

tharvik replied:
> I like the idea of having simple stateless functions for preprocessing. I would prefer if it was more organized, maybe by dataset type but that can be handled later on as we add more functions.

yep, it's a draft for now, it'll get nicer in the next PR. thanks for the remarks on small  functions, I think it indeed does help with understanding what is actually going on in preprocessing. it is also a requirement to have type safe preprocessing (as it modify the type of values which is sadly currently hidden via `tf.TensorContainer`)

> I think there may be some overlap between this file and `data/helpers.ts`, could you add some doc about the scope of each file to clarify where functions should go? On that matter maybe `convertors.ts` is a confusing name as not all functions there convert things (and some functions in helpers.ts also convert things).

yeah, the difference is not really clear in code. `data/helpers` is temporary and only for converting `Dataset` to `Data` & `DataSplit`, it'll get removed soon. `convertors` (or rather `processing` now) is a collections of functions to shape `Dataset` into what's models can ingest, on the way of replacing what the soon-to-be-old `data/preprocessing`.
I added some comments to make it clearer.

> What about something like `processing.ts` (for both pre- and post-processing)?

`processing` it is!

JulienVig commented on discojs/src/dataset/types.ts:
```diff
@@ -0,0 +1,12 @@
+import { Dataset } from "./dataset.js";
+
+// TODO use full type to check shape of array
```
What do you mean by "use full type"?

tharvik replied:
using a class in this case, to actively check that the array size is consistent with width, height and depth. it is now implemented in the latest commits.

JulienVig commented on discojs/src/models/model.ts:
```diff
@@ -19,6 +18,7 @@ export interface EpochLogs {
 // TODO still bound to tfjs
 export type Prediction = tf.Tensor;
 export type Sample = tf.Tensor;
+type Dataset = tf.data.Dataset<tf.TensorContainer>
```
I find the name overlap between the discojs Dataset and tf.data.Dataset confusing, what about renaming this type tfDataset or directly using tf.data.Dataset rather than creating a new type?
Or is it a temporary solution?

tharvik replied:
yeah, inline makes more sense, pushed a commit. I've been working on this PR for some time now, there is a bunch of leftovers from various temporary ideas

JulienVig commented on discojs/src/task/training_information.ts:
```diff
@@ -138,7 +138,7 @@ export function isTrainingInformation (raw: unknown): raw is TrainingInformation
       return false
     }
-    if (!(Array.isArray(outputColumns) && outputColumns.every((e) => typeof e === 'string'))) {
+    if (!(Array.isArray(outputColumn) && outputColumn.every((e) => typeof e === 'string'))) {
```
Doesn't `outputColumn` have to be a string now? Can it be an array?

tharvik replied:
ho right, that doesn't make sense, good catch!

JulienVig commented on webapp/src/components/data/TabularDatasetInput.vue:
There is now ImageDatasetInput and TabularDatasetInput, what about TextDatasetInput?

JulienVig commented on docs/examples/wikitext.ts:
Running `npm run language_model` in the `examples` folder now hangs forever (or a long time) between the training start and the first epoch. There's currently no issue on the develop branch.

tharvik replied:
yeah, that's bothering. it is working but very slowly as it is not using tfjs's node accelerator. adding it back in the example.

JulienVig commented on webapp/src/components/containers/ImageCard.vue:
```diff
@@ -1,43 +1,52 @@
       class="opacity-0 hover:opacity-100 duration-500 absolute inset-0 z-10 flex justify-center items-end text-white font-semibold p-5"
     >
-      <CustomButton
-        class="text-xs"
-      >
-        click to show on maps
-      </CustomButton>
+      <CustomButton class="text-xs"> click to show on maps </CustomButton>
     </div>
```
I think the showButton prop, the CustomButton and the draw function can be removed, they were probably used for the Polyfill task

tharvik replied:
indeed, `showButton` is never used, removing that.
I'm keeping the `draw` function as it is used when showing results of evaluation for images

JulienVig commented on webapp/src/components/data/ImageDatasetInput.vue:
```diff
@@ -0,0 +1,254 @@
+    );
+
+  if (!fullyFilled.value)
+    throw new Error("given folder is missing some files referenced in the CSV");
```
This error message is confusing because fullyFilled can be false if not all labels are represented in the dataset, even though all files referenced in the CSV are connected. This happened when I tried connecting datasets/cifar10-labels.csv

tharvik replied:
ha yes, I tried to only `emit` Dataset when theses are ~full, but you're right, one can input partial dataset also. removed the `fullyFilled` logic and inversed the folder parsing thus respecting the error message but doing it right :musical_note:

JulienVig reviewed (changes requested):
Huge amount of work, well done!! Things feel much cleaner now!

I think there are a few issues with the webapp right now:
* Clicking next on a task without connecting data doesn't display anthing anymore at the Training step (while it used to display the training board)
![Screenshot 2024-06-06 at 13 54 23](https://github.com/epfml/disco/assets/33122365/9b6b0fc3-a3d3-4742-9349-0e5543f3f8e8)
* Connecting data, clicking next to the Training board now displays the training things correctly but if I go back I can't clear the connected files anymore.
* Connecting the LUS COVID dataset and training fails with error `Error: Based on the provided shape, [224,224,3], the tensor should have 150528 values but has 200704`, maybe linked to the new image loading function? A similar errors occurs on simple face: `Error: Based on the provided shape, [200,200,3], the tensor should have 120000 values but has 160000`
* Training on titanic works well, however the webapp fails to display the testing page correctly afterward and the Test button doesn't appear. (Sidenote titanic training currently fails on develop with error: `Error: A Dataset iterator for fitDataset() is expected to generate objects of the form {xs: xVal, ys: yVal}, where the two values may be tf.Tensor, an array of Tensors, or a map of string to Tensor.  The provided Dataset instead generates Tensor` so you seemed to have fixed this error)
![Screenshot 2024-06-06 at 14 01 45](https://github.com/epfml/disco/assets/33122365/e7377b92-0a4a-438e-8a8b-85097858314a)
* Training on the Skin condition and CIFAR10 tasks fails with error `TypeError: valueScalar is undefined`, you can find the sample skin condition dataset [here](https://storage.googleapis.com/deai-313515.appspot.com/scin_sample.zip).
* There is no TextDatasetInput.vue so there is no way to connect data to the wikitext task in the webapp ![Screenshot 2024-06-06 at 14 10 57](https://github.com/epfml/disco/assets/33122365/aed05b39-5b6c-4545-b14f-a8724678efc6)

I've been terribly confused with the naming of `Dataset`, `Data`, `DataSplit` and `tf.data.Dataset`. A list of specific things I don't like about the current abstractions:
* Dataset is not a data structure composed of Data elements
* Data is actually a dataset (and it is `Data` that wraps `tf.data.Dataset` rather than `Dataset`)
* `DataSplit` represents a train-test split of `Data` rather than a dataset
* dataset ends up being an attribute of data like in `data.train.preprocess().batch().dataset`

Can we either add more class documentation to clarify the relationships or rework the oriented-object architecture in a way that makes things self-explanatory? This is a bit beyond the scope of this PR but it feels like the right time to address it? Feel free to ignore otherwise I'm also happy to create an additional PR myself.

tharvik committed:
discojs/Image: add full type

tharvik commented:
> * Clicking next on a task without connecting data doesn't display anthing anymore at the Training step (while it used to display the training board)

as no dataset was entered (it's undef), one can't train anything. I added a small message to tell users to enter some. one more functional way would be to rework the steps to disable the next step until data are entered.

> * Connecting data, clicking next to the Training board now displays the training things correctly but if I go back I can't clear the connected files anymore.

good catch! I indeed forgot to propagate the "clear file" event

> * Connecting the LUS COVID dataset and training fails with error `Error: Based on the provided shape, [224,224,3], the tensor should have 150528 values but has 200704`, maybe linked to the new image loading function? A similar errors occurs on simple face: `Error: Based on the provided shape, [200,200,3], the tensor should have 120000 values but has 160000`

indeed, I mixed RGBA & RGB images, added a real `Image` type and related `remove_alpha` convertor.

> * Training on titanic works well, however the webapp fails to display the testing page correctly afterward and the Test button doesn't appear.

right, thanks for catching that, I missplaced a `toRaw` in there. NB: this is needed as VueJS proxies every value, but it doesn't work with [JS private fields](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_properties) (`#` prefixed values).

> * Training on the Skin condition and CIFAR10 tasks fails with error `TypeError: valueScalar is undefined`, you can find the sample skin condition dataset [here](https://storage.googleapis.com/deai-313515.appspot.com/scin_sample.zip).

ho oupsi, I'm generating an empty dataset when uploading a folder, should have added a throwing TODO, thanks for catching that.

> * There is no TextDatasetInput.vue so there is no way to connect data to the wikitext task in the webapp

correct, there isn't any. we can add one in a next iteration but that out of scope of the PR. from what I remember, one student should have worked on that but it wasn't clear if it happened in the end.

> I've been terribly confused with the naming of `Dataset`, `Data`, `DataSplit` and `tf.data.Dataset`. A list of specific things I don't like about the current abstractions:

I totally agree with you, it's not understandable. that's also why this PR is introducing a new one that should replace all of theses in its next iterations (when I reworked the convertors).

> * Dataset is not a data structure composed of Data elements

that would be the central (and only) way to have a serie of data (image/tabular/text/...) with transformations applied on it.

> * Data is actually a dataset (and it is `Data` that wraps `tf.data.Dataset` rather than `Dataset`)

it will be dropped, replaced by `Dataset<Image>`, `Dataset<Tabular>` & `Dataset<Text>` and some convertors applied on it. 

> * `DataSplit` represents a train-test split of `Data` rather than a dataset

it will be replaced by the return type of `Dataset<T>.split()` ie `[Dataset<T>, Dataset<T>]`.

> * dataset ends up being an attribute of data like in `data.train.preprocess().batch().dataset`

it will be real values, not fields, by shaping `Dataset<T>` via convertors functions. for example, `Dataset<Tabular> -> extract_columns -> convert_to_numbers -> Dataset<[features: number[], label: number]>`

> Can we either add more class documentation to clarify the relationships or rework the oriented-object architecture in a way that makes things self-explanatory? This is a bit beyond the scope of this PR but it feels like the right time to address it? Feel free to ignore otherwise I'm also happy to create an additional PR myself.

as I'm hoping to remove most of theses soon, it seems a bit premature to update the doc now. I'll update the doc accordingly in the next one, but feel free to setup a PR if you feel that's it's taking too long.

JulienVig commented:
Sounds good! Is there a quick fix to support text dataset? Merging this PR means that wikitext will not be available on discolab.ai until a new PR implements text datasets.

JulienVig commented:
> as no dataset was entered (it's undef), one can't train anything. I added a small message to tell users to enter some. one more functional way would be to rework the steps to disable the next step until data are entered.

I quite liked simply displaying an error message when training without data ("input a dataset first"). That allowed new users to walk around and see things when they don't want to train an actual model. What do you think of it?

tharvik committed:
*: import tfjs-node for speed

tharvik commented:
> Is there a quick fix to support text dataset? Merging this PR means that wikitext will not be available on discolab.ai until a new PR implements text datasets.

hum, I think that's doable, not for the webapp Tester (it's not really supported now anyway) but for the others parts, yes.

> > as no dataset was entered (it's undef), one can't train anything. I added a small message to tell users to enter some. one more functional way would be to rework the steps to disable the next step until data are entered.
> 
> I quite liked simply displaying an error message when training without data ("input a dataset first"). That allowed new users to walk around and see things when they don't want to train an actual model. What do you think of it?

hum, users are weird, but yeah, make sense, it's very demo-y; adding it back.

tharvik commented:
so, it's alive again! way smaller to review thanks to extracted scraps. description updated.

JulienVig commented:
There seems to be something wrong with the image data loaders: connecting images (by group for lus_covid and by csv for mnist) results in an error (`Something went wrong on our side`) when clicking train (both alone and collaboratively). Titanic and the llm task work fine.

JulienVig commented on cli/src/data.ts:
```diff
@@ -1,76 +1,51 @@
+    case "simple_face":
+      return ["image", await loadSimpleFaceData()];
+    case "titanic":
+      return ["tabular", loadCSV("../datasets/titanic_train.csv")];
```
```suggestion
      return ["tabular", loadCSV(path.join("..", "datasets", "titanic_train.csv"))];
```
To make it consistent with line 42

JulienVig commented on discojs/src/dataset/dataset.ts:
```diff
@@ -1,6 +1,177 @@
+  | (() => AsyncIterator<T, void>)
+  | (() => Iterator<T, void>);
+
+/** Immutable serie of data */
```
```suggestion
/** Immutable series of data */
```
Series always has an s even when there's a single series...

tharvik replied:
human languages will never stop astonishing me :eye_speech_bubble:

JulienVig commented on discojs/src/dataset/image.ts:
```diff
@@ -0,0 +1,15 @@
+export class Image<
+  D extends 1 | 3 | 4 = 1 | 3 | 4,
```
Could you add a comment clarifying what the superclass is?

JulienVig commented on cli/src/benchmark_gpt.ts:
```diff
@@ -74,12 +87,29 @@ async function main(args: Required<CLIArguments>): Promise<void> {
+              .toArray(),
+          ) as tf.Tensor3D,
+        })),
+    );
```
I noticed a memory leak when running `DEBUG=* npm -w cli run benchmark_gpt`. I expect that the preprocessing needs some tf.tidy and tf.dispose

JulienVig replied:
Similarly, printing the peakMemory from the EpochLogs shows an increasing memory when using the webapp

tharvik replied:
yeah, I forgot to clean it nicely, it should be fixed now. (tfjs is hard to memory manage..)

tharvik committed:
webapp/tests: train lus covid

tharvik committed:
webapp/training: fix unraw dataset use

JulienVig commented on webapp/src/components/training/Training.vue:
```diff
@@ -13,106 +10,121 @@
-    />
+    <Finished v-show="trainingStore.step === 4" :task="task" />
+
+    <Finished v-show="trainingStore.step === 4" :task="task" />
```
```suggestion
```
Duplicate line

JulienVig commented on webapp/src/components/testing/Testing.vue:
```diff
@@ -165,152 +185,186 @@
+          `row ${i} is missing columns ${wantedColumns.subtract(columns).join(", ")}`,
+        );
+  } catch (e) {
+    let msg = "Error when loading CSV";
```
When connecting the same titanic csv I used during training at the testing step, I get the error: `Error when loading CSV: can't access private field or method: object is not the right class`

JulienVig replied:
Same error if I connect a misformatted csv

tharvik replied:
and same fix as before, was a bit fast on this one :sweat_smile:

JulienVig commented on webapp/src/components/testing/Testing.vue:
```diff
@@ -144,19 +165,18 @@
       <div v-if="stepRef === 1">
```
```suggestion
      <div 
        v-if="stepRef === 1"
        class="flex flex-col space-y-4 md:space-y-8"
      >
```
Follow same style as training. Right now there is no gap between the divs:
<img width="600" alt="Screenshot 2024-08-13 at 14 06 34" src="https://github.com/user-attachments/assets/a53fe5b2-da95-4f86-9ff6-cc101a9deba6">

JulienVig commented on webapp/src/components/dataset_input/LabeledImageDatasetInput/__tests__/ByCSV.spec.ts:
Why are we removing the tests?

JulienVig replied:
Ah I guess it's been moved to e2e/datasetInput.cy.ts

tharvik replied:
indeed; I moved to cypress as vue's test-utils is not moving fast [on v-model issues](https://github.com/vuejs/test-utils/issues/2468).

JulienVig commented on discojs/src/convertors.ts:
```diff
@@ -0,0 +1,126 @@
+ * @param length number of tokens
+ * @returns encoded string in an array of token, size of max_length
+ **/
+export function tokenizeAndLeftPad(
```
Right now this function is only used by the cli and not by the trainingInformation preprocessing, is this intened?

tharvik replied:
for now yes, I didn't wanted to modify too much the internals of discojs. a rewrite of preprocessing is baking in the next PR.

tharvik commented:
> There seems to be something wrong with the image data loaders

right, I missplaced a `toRaw` call. should be fixed. and added a test for lus_covid's training to catch this earlier.

JulienVig commented on docs/examples/training.ts:
```diff
@@ -66,35 +68,15 @@ async function main (): Promise<void> {
+  return adults.chain(childs);
 }
 
 // You can run this example with "npm start" from this folder
```
```suggestion
// You can run this example with "npm run train" from this folder
```

JulienVig commented on docs/examples/custom_task.ts:
```diff
@@ -1,4 +1,4 @@
-import tf from '@tensorflow/tfjs'
+import tf from '@tensorflow/tfjs-node'
```
This is probably due to something else than this import but running `npm run custom_task` hangs forever (or at least a long time)

tharvik replied:
hum, it's already the case in develop, is it not wanted? from a quick look, it seems to be only starting a server and never closing it (node doesn't exit as long a their are some stuff/promises running)

JulienVig reviewed (changes requested):
Impressive work! You've virtually rewritten most of the codebase haha

The evaluation seems to have an issue with data loading:
* Go to the Evaluation tab, test a GPT language model, connect a text file, click next: Toaster error "Something went wrong" and nothing displayed. Now If I click Previous the text file I previously connected has been cleared
* Same error with lus_covid (test a model, connect images, next -> nothing displayed).

tharvik committed:
cli: use path.join where possible

tharvik committed:
webapp/tests: commonize server setup

tharvik committed:
webapp/tests: add testing e2e

tharvik committed:
webapp/testing: fix text test

tharvik committed:
*: avoid some tf memory leak

tharvik committed:
discojs/prepocessing/text: fix type check

tharvik committed:
webapp/training: stop via AbortController

webapp/training: back to stop-by-throwing

This reverts commit 7b259af72b3724308bf9d4fed9ebf16227cc7652.

tharvik committed:
discojs: convertors -> processing

tharvik committed:
discojs: add some comments

tharvik committed:
examples/training: fix comment

tharvik committed:
webapp: small fixes

tharvik commented:
> Impressive work! You've virtually rewritten most of the codebase haha

haha, slowly getting there! when changing something as fundamental as tfjs, there is some good rewrite indeed.

> The evaluation seems to have an issue with data loading:

that should be fixed now, and some cypress tests to ensure that also. we still have a memory leak of one tensor every round. I didn't wanted to sink _too_ much time into it.

JulienVig reviewed (changes requested):
Thanks for the improvements!! I found a few things remaining:
* Training on lus_covid and then testing the model displays 0% accuracy and 0 samples visited
* Trying to test/predict a model on titanic data by connecting the titanic_test.csv or titanic_train.csv doesn't display enough and there is a toaster error
* Idem when trying to test a language model
* The Stop training button doesn't work when the training fails (for example if you train collaboratively with mnist with a single user, the user timeouts waiting for other contributions. After the timeout, hitting stop training doesn't work) 
* (At some point I also got the error "multiple dataset entered info=render function thrown undefined" but I can't reproduce it...)

tharvik committed:
discojs/validation: simplify

tharvik committed:
webapp/testing: specialize Tester

tharvik committed:
webapp/store/validation: simplify

tharvik commented:
> * Training on lus_covid and then testing the model displays 0% accuracy and 0 samples visited
> * Trying to test/predict a model on titanic data by connecting the titanic_test.csv or titanic_train.csv doesn't display enough and there is a toaster error
> * Idem when trying to test a language model

haa, one of the dark corner of disco. so I finally took hold of it and rework the whole discojs' validation stack
* predict yield the prediction for every line of the dataset
* test yield a boolean if the prediction is correct
* no more duplicated logic between the two, no more materialisation of the dataset's values in the results, no more mutability :sun_with_face:
* specialize Tester into PredictSteps & TestSteps
* simplify validation's store
  * we can even remove it down the road, if we rework the URL to actually encode the model ID in it

> * The Stop training button doesn't work when the training fails (for example if you train collaboratively with mnist with a single user, the user timeouts waiting for other contributions. After the timeout, hitting stop training doesn't work)

arf, too bad, it comes from the `AbortController` only reacting when between await points. and as Disco is never returning when an error occurs (that should be the case but I was already kinda yack shaving), it hangs on the `for await`. throwing allows for propagation of the failure. I'm reintroducing it and we can change it again when Disco behaves a bit more nicely.

> * (At some point I also got the error "multiple dataset entered info=render function thrown undefined" but I can't reproduce it...)

hum, that shouldn't happen. it comes when entering multiple type of dataset in test/predict, which shouldn't be possible. I tried a bit to reproduce it but wasn't able to.

JulienVig commented on webapp/src/store/validation.ts:
```diff
@@ -1,15 +1,9 @@
-  const model = ref<string>()
-  const isOnlyPrediction = ref(false)
+export const useValidationStore = defineStore("validation", () => {
+  const step = ref<0 | 1 | 2>(0);
```
Very neat!

JulienVig commented on webapp/src/components/testing/PredictSteps.vue:
```diff
@@ -0,0 +1,342 @@
+  >
+    <div class="flex justify-center items-center mb-4">
+      <span class="shrink-0 py-4 px-4 bg-orange-100 rounded-md">
+        <p class="text-slate-600 text-xs">
```
```suggestion
      <span class="shrink-0 py-4 px-4 bg-purple-100 rounded-md
        flex flex-row gap-x-4 items-center"
      >
        <InfoIcon custom-class="min-w-6 min-h-6 w-6 h-6 text-slate-600" />
```
To follow the new purple banner style
Make sure to import InfoIcon :)
```
import InfoIcon from "@/assets/svg/InfoIcon.vue";
```

JulienVig replied:
hmm it seems the code suggestion isn't pointing at the right line... it should point to l.8 not l.9

JulienVig reviewed (approved):
Awesome!! Better validator and validator store and the test suite is getting more and more significant!

All good for me, could you maybe merge your PR after #710 (or rebase on it) so that @tomasoignons doesn't have to do it?

martinjaggi commented:
huge thanks, massive change here!

JulienVig commented:
Yay we've done it! 🎉

martinjaggi commented:
congrats!
